File: src/calcio_manager/engine/commentary.py
"""Locale-aware commentary engine for match events."""

from __future__ import annotations

import random
import tomllib
from pathlib import Path
from uuid import UUID

from calcio_manager.i18n import current_locale
from calcio_manager.models.enums import MatchEventType
from calcio_manager.models.match import MatchEvent
from calcio_manager.models.player import Player
from calcio_manager.models.team import Team

DATA_DIR = Path(__file__).parent.parent / "data"

# Map event types to TOML section names
_EVENT_SECTION_MAP: dict[MatchEventType, str] = {
    MatchEventType.GOAL: "goal",
    MatchEventType.CHANCE: "chance",
    MatchEventType.SHOT_SAVED: "shot_saved",
    MatchEventType.SHOT_WIDE: "shot_wide",
    MatchEventType.SHOT_POST: "shot_post",
    MatchEventType.FOUL: "foul",
    MatchEventType.YELLOW_CARD: "yellow_card",
    MatchEventType.RED_CARD: "red_card",
    MatchEventType.SECOND_YELLOW: "second_yellow",
    MatchEventType.CORNER: "corner",
    MatchEventType.KICK_OFF: "kick_off",
    MatchEventType.HALF_TIME: "half_time",
    MatchEventType.FULL_TIME: "full_time",
    MatchEventType.PENALTY_SHOOTOUT_START: "penalty_shootout_start",
    MatchEventType.PENALTY_SCORED: "penalty_scored",
    MatchEventType.PENALTY_MISSED: "penalty_missed",
    MatchEventType.PENALTY_SAVED: "penalty_saved",
    MatchEventType.POSSESSION: "possession",
    MatchEventType.TIMEOUT: "timeout",
    MatchEventType.SUBSTITUTION: "substitution",
}


class CommentaryEngine:
    """Generates text commentary for match events in the current locale."""

    def __init__(self) -> None:
        self._templates: dict[str, list[str]] = {}
        self._load_templates()

    def _load_templates(self) -> None:
        """Load commentary templates from the locale-specific TOML file."""
        lang = current_locale()
        path = DATA_DIR / f"commentary_{lang}.toml"
        if not path.exists():
            path = DATA_DIR / "commentary_it.toml"
        with open(path, "rb") as f:
            data = tomllib.load(f)

        for section_name, section_data in data.items():
            if isinstance(section_data, dict) and "templates" in section_data:
                self._templates[section_name] = section_data["templates"]
                # Also load special template variants
                for key, value in section_data.items():
                    if key != "templates" and isinstance(value, list):
                        self._templates[f"{section_name}_{key}"] = value

    def generate(
        self,
        event: MatchEvent,
        home_team: Team,
        away_team: Team,
        home_goals: int = 0,
        away_goals: int = 0,
        players: dict[UUID, Player] | None = None,
    ) -> str:
        """Generate commentary text for a match event.

        Args:
            event: The match event to commentate.
            home_team: Home team.
            away_team: Away team.
            home_goals: Current home goals.
            away_goals: Current away goals.
            players: Optional player lookup dict.
        """
        section = _EVENT_SECTION_MAP.get(event.event_type)
        if section is None:
            return ""

        # Build template variables
        team = home_team if event.team_id == home_team.id else away_team
        opponent = away_team if event.team_id == home_team.id else home_team

        player_name = ""
        if event.player_id and players:
            player = players.get(event.player_id)
            if player:
                player_name = player.short_name

        assist_name = ""
        if event.assist_player_id and players:
            assist_player = players.get(event.assist_player_id)
            if assist_player:
                assist_name = assist_player.short_name

        score = f"{home_goals} - {away_goals}"

        vars_dict = {
            "team": team.name,
            "opponent": opponent.name,
            "player": player_name,
            "scorer": player_name,
            "assist": assist_name,
            "home": home_team.name,
            "away": away_team.name,
            "home_goals": str(home_goals),
            "away_goals": str(away_goals),
            "score": score,
        }

        # Choose template variant
        template_key = section
        if event.event_type == MatchEventType.GOAL and assist_name:
            # Prefer assist variant if available
            assist_key = f"{section}_templates_with_assist"
            if assist_key in self._templates:
                template_key = assist_key

        templates = self._templates.get(template_key, [])
        if not templates:
            templates = self._templates.get(section, [])

        if not templates:
            return f"{event.minute}' - {event.event_type.value}"

        template = random.choice(templates)

        try:
            return template.format(**vars_dict)
        except KeyError:
            return template

File: src-tauri/src/domain/engine/commentary.rs
use std::collections::HashMap;
use std::sync::OnceLock;

use toml::Value;

use crate::domain::models::MatchEventType;

const COMMENTARY_IT: &str = include_str!("../../../resources/data/commentary_it.toml");

type TemplateMap = HashMap<String, Vec<String>>;

#[derive(Debug, Clone, Default)]
pub struct CommentaryVars {
    pub team: String,
    pub opponent: String,
    pub player: String,
    pub scorer: String,
    pub assist: String,
    pub home: String,
    pub away: String,
    pub home_goals: u32,
    pub away_goals: u32,
}

impl CommentaryVars {
    fn score(&self) -> String {
        format!("{} - {}", self.home_goals, self.away_goals)
    }
}

fn templates() -> &'static TemplateMap {
    static TEMPLATES: OnceLock<TemplateMap> = OnceLock::new();
    TEMPLATES.get_or_init(load_templates)
}

fn load_templates() -> TemplateMap {
    let parsed = COMMENTARY_IT
        .parse::<Value>()
        .unwrap_or(Value::Table(Default::default()));
    let mut map = HashMap::new();

    let Some(table) = parsed.as_table() else {
        return map;
    };

    for (section, value) in table {
        let Some(section_table) = value.as_table() else {
            continue;
        };
        for (key, value) in section_table {
            let Some(items) = value.as_array() else {
                continue;
            };
            let templates: Vec<String> = items
                .iter()
                .filter_map(|item| item.as_str().map(ToString::to_string))
                .collect();
            if templates.is_empty() {
                continue;
            }
            let map_key = if key == "templates" {
                section.clone()
            } else {
                format!("{}_{}", section, key)
            };
            map.insert(map_key, templates);
        }
    }

    map
}

pub fn generate_commentary(
    event_type: MatchEventType,
    vars: &CommentaryVars,
    minute: u32,
) -> String {
    let Some(section) = section_for_event(event_type) else {
        return fallback(event_type, vars);
    };

    let key = if event_type == MatchEventType::Gol && !vars.assist.is_empty() {
        "goal_templates_with_assist"
    } else {
        section
    };

    let selected = templates()
        .get(key)
        .or_else(|| templates().get(section))
        .and_then(|items| {
            if items.is_empty() {
                None
            } else {
                let index = deterministic_index(items.len(), minute, event_type, vars);
                items.get(index)
            }
        });

    match selected {
        Some(template) => interpolate(template, vars),
        None => fallback(event_type, vars),
    }
}

fn deterministic_index(
    len: usize,
    minute: u32,
    event_type: MatchEventType,
    vars: &CommentaryVars,
) -> usize {
    let mut hash = minute as usize ^ event_type as usize;
    for byte in vars
        .team
        .bytes()
        .chain(vars.player.bytes())
        .chain(vars.assist.bytes())
    {
        hash = hash.wrapping_mul(31).wrapping_add(byte as usize);
    }
    hash % len
}

fn section_for_event(event_type: MatchEventType) -> Option<&'static str> {
    match event_type {
        MatchEventType::Gol => Some("goal"),
        MatchEventType::Occasione => Some("chance"),
        MatchEventType::Parata => Some("shot_saved"),
        MatchEventType::TiroFuori => Some("shot_wide"),
        MatchEventType::Palo => Some("shot_post"),
        MatchEventType::Fallo => Some("foul"),
        MatchEventType::Ammonizione => Some("yellow_card"),
        MatchEventType::Espulsione => Some("red_card"),
        MatchEventType::DoppiaAmmonizione => Some("second_yellow"),
        MatchEventType::Sostituzione => Some("substitution"),
        MatchEventType::CalcioAngolo => Some("corner"),
        MatchEventType::Intervallo => Some("half_time"),
        MatchEventType::FinePartita => Some("full_time"),
        MatchEventType::CalcioInizio => Some("kick_off"),
        MatchEventType::InizioRigori => Some("penalty_shootout_start"),
        MatchEventType::RigoreSegnato => Some("penalty_scored"),
        MatchEventType::RigoreSbagliato => Some("penalty_missed"),
        MatchEventType::RigoreParato => Some("penalty_saved"),
        MatchEventType::Possesso => Some("possession"),
        MatchEventType::TimeOut => Some("timeout"),
        MatchEventType::Punizione => Some("free_kick"),
    }
}

fn interpolate(template: &str, vars: &CommentaryVars) -> String {
    template
        .replace("{team}", &vars.team)
        .replace("{opponent}", &vars.opponent)
        .replace("{player}", safe_value(&vars.player, "un giocatore"))
        .replace("{scorer}", safe_value(&vars.scorer, "un giocatore"))
        .replace("{assist}", safe_value(&vars.assist, "un compagno"))
        .replace("{home}", &vars.home)
        .replace("{away}", &vars.away)
        .replace("{home_goals}", &vars.home_goals.to_string())
        .replace("{away_goals}", &vars.away_goals.to_string())
        .replace("{score}", &vars.score())
}

fn safe_value<'a>(value: &'a str, fallback: &'a str) -> &'a str {
    if value.is_empty() { fallback } else { value }
}

fn fallback(event_type: MatchEventType, vars: &CommentaryVars) -> String {
    match event_type {
        MatchEventType::Gol => format!(
            "GOOOL! {} segna per {}!",
            safe_value(&vars.scorer, "un giocatore"),
            vars.team
        ),
        MatchEventType::Intervallo => format!(
            "Intervallo: {} {} - {} {}",
            vars.home, vars.home_goals, vars.away_goals, vars.away
        ),
        MatchEventType::FinePartita => format!(
            "Fine partita: {} {} - {} {}",
            vars.home, vars.home_goals, vars.away_goals, vars.away
        ),
        MatchEventType::CalcioInizio => format!("Si gioca! {} contro {}.", vars.home, vars.away),
        _ => format!("Evento: {:?}", event_type),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn loads_goal_templates() {
        assert!(
            templates()
                .get("goal")
                .is_some_and(|items| !items.is_empty())
        );
    }

    #[test]
    fn interpolates_commentary() {
        let vars = CommentaryVars {
            team: "Aurora".to_string(),
            player: "M. Rossi".to_string(),
            scorer: "M. Rossi".to_string(),
            home: "Aurora".to_string(),
            away: "Stella".to_string(),
            home_goals: 1,
            away_goals: 0,
            ..Default::default()
        };
        let text = generate_commentary(MatchEventType::Gol, &vars, 12);
        assert!(text.contains("Rossi") || text.contains("GOOOL"));
    }
}
